pkg/controller/job/job_controller.go

```diff
@@ -902,20 +902,30 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 
 	var manageJobErr error
 
+	// This is the starting point for evaluating the end state of the Job.
+	// Note that we need to order evaluations since a Job could satisfy multiple criteria at the same time in some cases:
+	// 1. Evaluate the pre-existing SuccessCriteriaMet and FailureTarget to respect the previous reconcile results.
+	// 2. Evaluate failure scenarios
+	// 3. Evaluate success scenarios
+
 	exceedsBackoffLimit := jobCtx.failed > *job.Spec.BackoffLimit
+	// Evaluate the pre-existing SuccessCriteriaMet.
 	jobCtx.finishedCondition = hasSuccessCriteriaMetCondition(&job)
 
 	// Given that the Job already has the SuccessCriteriaMet condition, the termination condition already had confirmed in another cycle.
 	// So, the job-controller evaluates the podFailurePolicy only when the Job doesn't have the SuccessCriteriaMet condition.
 	if jobCtx.finishedCondition == nil {
+		// Evaluate the pre-existing FailureTarget.
 		failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget)
 		if failureTargetCondition != nil && failureTargetCondition.Status == v1.ConditionTrue {
 			jobCtx.finishedCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
+			// Evaluate failure scenarios for the PodFailurePolicy.
```

Comment:
```suggestion
			// Evaluate failure scenarios for PodFailurePolicy.
```
Disclaimer: not an English expert here, but I think PodFailurePolicy is a proper noun in this context, so we can skip "the". Same for other feature names below.

Reply:
I'm also not well familiar with English :) So, I have no preference or opinion on this.

Reply:
in that case, maybe apply the suggestion, but feel free to change back if others request.

Reply:
Sure, makes sense.
Done.